[FLINK-25426][runtime] Calling super#onStop for all RpcEndpoint children to be sure that all resources close correctly

akalash opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*This PR fixes stopping of RpcEndpoint#MainThreadExecutor*


## Brief change log

  - *Call the RpcEndpoint#onStop in right places*
  - *Allowing to call MainThreadExecutor#execute at any time even after shutdown of scheduledExecutorService*


## Verifying this change

Existing UnalignedCheckpointRescaleITCase.

This change added tests and can be verified as follows:

  - *Added new test to RpcEndpointTest*
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

akalash commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java:
```diff
@@ -427,13 +427,7 @@ public void validateRunsInMainThread() {
 
         @Override
         public void execute(@Nonnull Runnable command) {
-            if (!scheduledExecutorService.isShutdown()) {
```
We can not ignore even one runnable here since MainThreadExecutor can be the executor for CompleatableFuture and if scheduledExecutorService would be shutdown before all futures will be done we can lost important execution like releasing resources. For example, take a look at org.apache.flink.runtime.scheduler.SchedulerBase#closeAsync where shutDownCheckpointServices is called exactly as I described before.

KarmaGYZ replied:
You are right. @zjureel I think we need to do the same investigation for `schedule` as well.

akalash commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -275,7 +275,7 @@ private void handleStartDispatcherServicesException(Exception e) throws Exceptio
     public CompletableFuture<Void> onStop() {
         log.info("Stopping dispatcher {}.", getAddress());
-
+        super.onStop();
```
Honestly, I don't like this solution since it looks pretty fragile. I tried to move closing mainThreadExecutor to `RpcEndpoint#internalCallOnStop` which I like more but unfortunately, it is impossible to do so because we have one more `Executor` in `FencedRpcEndpoint`.
In my opinion, we should rethink the management of resources here. Perhaps we should register all resources in one `CloseableRegistry` and then close it in one place(for example, in `RpcEndpoint#internalCallOnStop` after `onStop` future)

zjureel replied:
Thanks @akalash. I like the idea of `ClosableRegistry` to close all resources in one place. In fact it confuses me a little while there're `stop`, `onStop`, `internalCallOnStop` and `closeAsync` in `RpcEndpoint`. I think it's very good and necessary to close all the resources in one place

KarmaGYZ replied:
Actually, I don't have a strong preference here. In the current implementation, we need to claim in the JavaDoc of `RpcEndpoint#onStop` that the `super.onStop()` should always be called.
I agree it looks fragile and `AutoCloseableRegistry` sounds like a good idea. However, one can also complain that people might forget to register the resources to the `AutoCloseableRegistry`.
Either way, I think it is important to document it in a prominent position.

akalash replied:
I thought about it a little more. Right now, I think that in general, we need to have one pattern for handling resources but it is not clear what it should be for now and I agree that `AutoCloseableRegistry` also has a problem. So maybe it makes sense to leave the changes as is at least for now.
If nobody minds of these changes let's wait for the test result and then merge it. @KarmaGYZ

KarmaGYZ replied:
Sounds good to me. Would you like to file a ticket for the rethink of resource cleanup protocol in `RpcEndpoint`?

akalash replied:
Perhaps, it makes sense. I need to think about how to formulate the problem(Highly likely, it is not only about `RpcEndpoint` but for the similar classes).

akalash commented:
I have updated a little javadoc for RpcEndpoint#onStop

KarmaGYZ commented:
@akalash Please take a look at the failed tests.

akalash commented:
@KarmaGYZ, thanks for the notification, I have updated the PR. Tests are green now. Please, take a final look.

akalash commented:
I have added one more commit with enabling SavepointITCase#testTriggerSavepointAndResumeWithNoClaim since I think it should be also fixed by this fix. Original task for this test - https://issues.apache.org/jira/browse/FLINK-25427

akalash commented:
@KarmaGYZ, the PR is fully ready to merge, so please take a look at it. As I understand there are plenty tests still fail in master so it will be great to merge it today.

twalthr commented:
FYI: see also the discussion in FLINK-18356

akalash commented:
The PR is closed since the problematic commit was reverted. The idea behind this PR can be used for the proper fix of the original ticket.
